cardano-node/src/Cardano/Node/Tracing/Tracers/BlockReplayProgress.hs

```diff
@@ -71,28 +88,34 @@ instance MetaTrace ReplayBlockStats where
 
 withReplayedBlock :: Trace IO ReplayBlockStats
     -> IO (Trace IO (ChainDB.TraceEvent blk))
-withReplayedBlock tr =
-    let tr' = filterTrace filterFunction tr
-        tr'' = contramap unfold tr'
-    in foldTraceM replayBlockStats emptyReplayBlockStats tr''
+withReplayedBlock tr = do
+  var <- newMVar initialReplayBlockState
+  contramapMCond tr (process var)
   where
-    filterFunction(_, ReplayBlockStats {..}) = rpsDisplay
-
-replayBlockStats :: MonadIO m
-  => ReplayBlockStats
-  -> LoggingContext
-  -> ChainDB.TraceEvent blk
-  -> m ReplayBlockStats
-replayBlockStats ReplayBlockStats {..} _context
-    (ChainDB.TraceLedgerDBEvent
-     (LedgerDB.LedgerReplayEvent
-      (LedgerDB.TraceReplayProgressEvent
-       (LedgerDB.ReplayedBlock pt [] _ (LedgerDB.ReplayGoal replayTo))))) = do
-      let slotno = realPointSlot pt
-          endslot = withOrigin 0 id $ pointSlot replayTo
-          progress' = (fromIntegral (unSlotNo slotno) * 100.0) / fromIntegral (unSlotNo $ max slotno endslot)
-      pure $ if (progress' == 0.0 && not rpsDisplay)
-                || ((progress' - rpsLastProgress) > 0.1)
-                then ReplayBlockStats True slotno endslot progress' progress'
-                else ReplayBlockStats False slotno endslot progress' rpsLastProgress
-replayBlockStats st@ReplayBlockStats {} _context _ = pure st
+    process :: MVar ReplayBlockState
+            -> (LoggingContext, Either TraceControl (ChainDB.TraceEvent blk))
+            -> IO (Maybe (LoggingContext, Either TraceControl ReplayBlockStats))
+    process _ (ctx, Left control) = pure (Just (ctx, Left control))
+    process var (ctx, Right msg) = modifyMVar var $ \st -> do
```

Comment:
A *very* minor detail: As per the Haskell2010 report "A case expression is evaluated by pattern matching the expression e against the individual alternatives. The alternatives are tried sequentially, from top to bottom."

For block replay it's not much of an issue, but in general for the tracing system it's advisable to match on the `Right` case first (as it's done all throughout the rest of similar contexts).